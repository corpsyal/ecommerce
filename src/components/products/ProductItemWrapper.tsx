import React from 'react';
import Card from '@material-ui/core/Card';
import CardActionArea from '@material-ui/core/CardActionArea';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import CardMedia from '@material-ui/core/CardMedia';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core';

const useStyles = makeStyles({
  root: {
    maxWidth: 345,
  },
  media: {
    height: 280,
    objectFit: 'scale-down'
  },
  price: {
    display: 'block',
    paddingTop: 24,
    textAlign: 'center'
  },
  link: {
    width: '100%',
    textDecoration: 'none'
  }
});

interface IProductItemWrapperProps {
  product: IProduct,
  children: React.ReactNode
}

const ProductItemWrapper = ({ product, children }: IProductItemWrapperProps) => {
  const classes = useStyles();

  return (
    <Card className={classes.root}>
      <CardActionArea>
        <CardMedia
          component="img"
          alt="Contemplative Reptile"
          height="140"
          image={product.pictures[0]}
          title="Contemplative Reptile"
          className={classes.media}
        />
        <CardContent>
          <Typography gutterBottom variant="h6" component="h6">
            {product.name}
          </Typography>
          <Typography variant="body2" color="textSecondary" component="p">
            {product.description}
          </Typography>
          <Typography variant="h4" component="span" className={classes.price}>
            {product.price}â‚¬
              </Typography>
        </CardContent>
      </CardActionArea>
      <CardActions>
          {children}
      </CardActions>
    </Card>
  );
}

export default ProductItemWrapper;