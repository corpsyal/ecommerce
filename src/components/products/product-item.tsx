import React from 'react';
import Card from '@material-ui/core/Card';
import CardActionArea from '@material-ui/core/CardActionArea';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import CardMedia from '@material-ui/core/CardMedia';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core';
import { Link } from 'react-router-dom';

const useStyles = makeStyles({
  root: {
    maxWidth: 345,
  },
  media: {
    height: 280,
    objectFit: 'scale-down'
  },
  price: {
    display: 'block',
    paddingTop: 24,
    textAlign: 'center'
  },
  link: {
    width: '100%',
    textDecoration: 'none'
  }
});

interface IProductItemProps {
  product: IProduct
}

const ProductItem = ({ product }: IProductItemProps) => {
  const classes = useStyles();

  return (
    <Card className={classes.root}>
      <CardActionArea>
        <CardMedia
          component="img"
          alt="Contemplative Reptile"
          height="140"
          image={product.pictures[0]}
          title="Contemplative Reptile"
          className={classes.media}
        />
        <CardContent>
          <Typography gutterBottom variant="h6" component="h6">
            {product.name}
          </Typography>
          <Typography variant="body2" color="textSecondary" component="p">
            {product.description}
          </Typography>
          <Typography variant="h4" component="span" className={classes.price}>
            {product.price}â‚¬
              </Typography>
        </CardContent>
      </CardActionArea>
      <CardActions>
        {/* Button tag in a Link tag is not valid html, it's just for sample, in real usecase we must take the time to style the link component */}
        <Link to={`/${product.id}`} className={classes.link}>
          <Button size="large" variant="contained" color="primary" fullWidth>
            Voir
          </Button>
        </Link>

      </CardActions>
    </Card>
  );
}

export default ProductItem;