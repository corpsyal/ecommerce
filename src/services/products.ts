export const products: IProduct[] = [
    {
        id: '9myf5qjyvr',
        name: 'TISSOT CHEMIN DES TOURELLES POWERMATIC 80',
        price: 960,
        pictures: [
            "https://www.tissotwatches.com/media/shop/catalog/product/cache/all/image/920x/17f82f742ffe127f42dca9de82fb58b1/T/0/T099.429.11.038.00_1.png"
        ],
        description: "La Tissot Gentleman est une montre polyvalente, à la fois ergonomique et élégante en toutes circonstances. Elle se porte aussi bien en environnement professionnel, qui privilégie des codes classiques, que le week-end, où elle se plie avec aisance aux activités de loisirs. A sa place dans la vie d'un homme moderne et actif, la Tissot Gentleman se fait le compagnon idéal de tous les jours, de toutes les occasions, de tous les styles."
    },
    {
        id: 'zlsnz1xlzw',
        name: 'TISSOT CARSON PREMIUM POWERMATIC 80',
        price: 570,
        pictures: [
            "https://www.tissotwatches.com/media/shop/catalog/product/cache/all/image/920x/17f82f742ffe127f42dca9de82fb58b1/T/1/T122.407.16.031.00_1.png"
        ],
        description: "La Tissot Gentleman est une montre polyvalente, à la fois ergonomique et élégante en toutes circonstances. Elle se porte aussi bien en environnement professionnel, qui privilégie des codes classiques, que le week-end, où elle se plie avec aisance aux activités de loisirs. A sa place dans la vie d'un homme moderne et actif, la Tissot Gentleman se fait le compagnon idéal de tous les jours, de toutes les occasions, de tous les styles."
    },
    {
        id: 'd23d6k2wbb',
        name: 'TISSOT GENTLEMAN POWERMATIC 80 SILICIUM',
        price: 750,
        pictures: [
            "https://www.tissotwatches.com/media/shop/catalog/product/cache/all/image/920x/17f82f742ffe127f42dca9de82fb58b1/T/1/T127.407.16.041.00.png"
        ],
        description: "La Tissot Gentleman est une montre polyvalente, à la fois ergonomique et élégante en toutes circonstances. Elle se porte aussi bien en environnement professionnel, qui privilégie des codes classiques, que le week-end, où elle se plie avec aisance aux activités de loisirs. A sa place dans la vie d'un homme moderne et actif, la Tissot Gentleman se fait le compagnon idéal de tous les jours, de toutes les occasions, de tous les styles."
    }
];

export const findProductById = (products: IProduct[] = []) => (id: string): IProduct | undefined => products.find(product => product.id === id)
export const calcTotalPrice = (products: IProduct[] = []) => products.reduce((totalPrice: number, product: IProduct) => product.price + totalPrice , 0)

const delayDuration: number = 3000; // in milliseconds
 
type FetchAllProducts = () => Promise<IProduct[]>;
export const fetchAllProducts: FetchAllProducts = () => new Promise(async resolve => {
    setTimeout(() => resolve(products), delayDuration); // to simulate an api call
});

type FetchProductById = (id: string) => Promise<IProduct>;
export const fetchProductById: FetchProductById = id => new Promise(async resolve => {
    setTimeout(() => {
        const product = findProductById(products)(id);
        resolve(product);
    }, delayDuration);
});

