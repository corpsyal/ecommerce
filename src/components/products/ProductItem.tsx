import React from 'react';
import Button from '@material-ui/core/Button';
import { makeStyles } from '@material-ui/core';
import { Link } from 'react-router-dom';
import ProductItemWrapper from './ProductItemWrapper';

const useStyles = makeStyles({
  link: {
    width: '100%',
    textDecoration: 'none'
  }
});

interface IProductItemProps {
  product: IProduct
}

const ProductItem = ({ product }: IProductItemProps) => {
  const classes = useStyles();

  return (
    <ProductItemWrapper product={product}>
      {/* Button tag in a Link tag is not valid html, it's just for sample, in real usecase we must take the time to style the link component */}
      <Link to={`/${product.id}`} className={classes.link}>
          <Button size="large" variant="contained" color="primary" fullWidth>
            Voir
          </Button>
        </Link>
    </ProductItemWrapper>
  );
}

export default ProductItem;