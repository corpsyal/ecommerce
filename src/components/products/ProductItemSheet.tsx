import React from 'react';
import Button from '@material-ui/core/Button';
import ProductItemWrapper from './ProductItemWrapper';

interface IProductItemSheetProps {
  product: IProduct
}

const ProductItemSheet = ({ product }: IProductItemSheetProps) => {

  return (
    <ProductItemWrapper product={product}>
      {/* Button tag in a Link tag is not valid html, it's just for sample, in real usecase we must take the time to style the link component */}
          <Button size="large" variant="contained" color="primary" fullWidth>
            Ajouter au panier
          </Button>
    </ProductItemWrapper>
  );
}

export default ProductItemSheet;