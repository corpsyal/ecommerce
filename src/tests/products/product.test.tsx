import React from "react";
import { shallow } from "enzyme";
import ProductItem from "../../components/products/product-item";
import { products } from "../../services/products";

describe('ProductItem rendering', () => {
    it('Should render correctly', () => {
        const wrapper = shallow(<ProductItem product={products[0]} />);
        expect(wrapper).toMatchSnapshot();
    });
});