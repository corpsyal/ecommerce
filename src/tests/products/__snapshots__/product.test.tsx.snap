// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductItem rendering Should render correctly 1`] = `
<WithStyles(ForwardRef(Card))
  className="makeStyles-root-1"
>
  <WithStyles(ForwardRef(CardActionArea))>
    <WithStyles(ForwardRef(CardMedia))
      alt="Contemplative Reptile"
      className="makeStyles-media-2"
      component="img"
      height="140"
      image="https://www.tissotwatches.com/media/shop/catalog/product/cache/all/image/920x/17f82f742ffe127f42dca9de82fb58b1/T/0/T099.429.11.038.00_1.png"
      title="Contemplative Reptile"
    />
    <WithStyles(ForwardRef(CardContent))>
      <WithStyles(ForwardRef(Typography))
        component="h6"
        gutterBottom={true}
        variant="h6"
      >
        TISSOT CHEMIN DES TOURELLES POWERMATIC 80
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        color="textSecondary"
        component="p"
        variant="body2"
      >
        La Tissot Gentleman est une montre polyvalente, à la fois ergonomique et élégante en toutes circonstances. Elle se porte aussi bien en environnement professionnel, qui privilégie des codes classiques, que le week-end, où elle se plie avec aisance aux activités de loisirs. A sa place dans la vie d'un homme moderne et actif, la Tissot Gentleman se fait le compagnon idéal de tous les jours, de toutes les occasions, de tous les styles.
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-price-3"
        component="span"
        variant="h4"
      >
        960
        €
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(CardContent))>
  </WithStyles(ForwardRef(CardActionArea))>
  <WithStyles(ForwardRef(CardActions))>
    <Link
      className="makeStyles-link-4"
      to="/9myf5qjyvr"
    >
      <WithStyles(ForwardRef(Button))
        color="primary"
        fullWidth={true}
        size="large"
        variant="contained"
      >
        Voir
      </WithStyles(ForwardRef(Button))>
    </Link>
  </WithStyles(ForwardRef(CardActions))>
</WithStyles(ForwardRef(Card))>
`;
